File: src/main/java/org/kite9/diagram/common/objects/Bounds.java
package org.kite9.diagram.common.objects;

import org.apache.commons.math.fraction.BigFraction;

public interface Bounds extends Comparable<Bounds> {

	public abstract double getDistanceMin();

	public abstract double getDistanceMax();

	public abstract double getDistanceCenter();
	
	/** 
	 * Returns a new bounds large enough to take both this and other.
	 */
	public Bounds expand(Bounds other);
	
	/**
	 * Returns a new bounds with just the common distance inside
	 */
	public Bounds narrow(Bounds other);
	
	public Bounds keep(double buffer, double width, BigFraction atFraction);
	
	public Bounds keep(double buffer, double width, double atFraction);
	
//	public Bounds keepMax(double lb, double ub);
//	
//	public Bounds keepMin(double ub, double lb);
//	
//	public Bounds keepMid(double w);

}

File: src/main/kotlin/org/kite9/diagram/common/objects/Bounds.kt
package org.kite9.diagram.common.objects

import org.kite9.diagram.common.fraction.BigFraction

interface Bounds : Comparable<Bounds> {
    val distanceMin: Double
    val distanceMax: Double
    val distanceCenter: Double

    /**
     * Returns a new bounds large enough to take both this and other.
     */
    fun expand(other: Bounds): Bounds?

    /**
     * Returns a new bounds with just the common distance inside
     */
    fun narrow(other: Bounds): Bounds
    fun keep(buffer: Double, width: Double, atFraction: BigFraction): Bounds
    fun keep(buffer: Double, width: Double, atFraction: Double): Bounds
    fun narrow(trim: Double): Bounds

}